lib/src/annotate.rs

```diff
@@ -0,0 +1,421 @@
+        if is_first {
+            results.swap_full_commit_id(starting_commit_id, &cid);
+            is_first = false;
+        }
+        process_commit(results, repo, file_name, &cid, &edge_list)?;
+        if results.original_line_map.len() >= num_lines {
+            break;
+        }
+    }
+    Ok(())
+}
+
+/// For a given commit, for each parent, we compare the version in the parent
+/// tree with the current version, updating the mappings for any lines in
+/// common. If the parent doesn't have the file, we skip it.
+/// After iterating through all the parents, any leftover lines unmapped means
+/// that those lines are original in the current commit. In that case,
+/// original_line_map is updated for the leftover lines.
+fn process_commit(
+    results: &mut PartialResults,
+    repo: &ReadonlyRepo,
+    file_name: &RepoPath,
+    current_commit_id: &CommitId,
+    edges: &Vec<GraphEdge<CommitId>>,
+) -> Result<(), AnnotateError> {
+    let current_tree = repo.store().get_commit(current_commit_id)?.tree()?;
+    let current_file_id = get_file_id(current_commit_id, &current_tree, file_name)?.unwrap();
+
+    for parent_edge in edges {
+        if parent_edge.edge_type != GraphEdgeType::Missing {
+            let parent_commit = repo.store().get_commit(&parent_edge.target)?;
+            let parent_tree = parent_commit.tree()?;
+            let parent_file_id = get_file_id(parent_commit.id(), &parent_tree, file_name)?;
+
+            if let Some(pfid) = parent_file_id {
+                process_file_ids(
+                    results,
+                    repo.store(),
+                    file_name,
+                    &current_file_id,
+                    current_commit_id,
+                    &pfid,
+                    parent_commit.id(),
+                )?;
+            }
+        }
+    }
+    results.drain_remaining_for_commit_id(current_commit_id);
+
+    Ok(())
+}
+
+/// For two versions of the same file, for all the lines in common, overwrite
+/// the new mapping in the results for the new commit. Meaning, Let's say I have
+/// a file in commit A and commit B. We know that according to local_line_map,
+/// in commit A, line 3 corresponds to line 7 of the original file. Now, line 3
+/// in Commit A corresponds to line 6 in commit B. Then, we update
+/// local_line_map to say that "Commit B line 6 goes to line 7 of the original
+/// file". We repeat this for all lines in common in the two commits. For 2
+/// identical files, we bulk replace all mappings from commit A to commit B in
+/// local_line_map
+fn process_file_ids(
+    results: &mut PartialResults,
+    store: &Store,
+    file_name: &RepoPath,
+    current_file_id: &FileId,
+    current_commit_id: &CommitId,
+    parent_file_id: &FileId,
+    parent_commit_id: &CommitId,
+) -> BackendResult<()> {
+    if current_file_id == parent_file_id {
+        results.swap_full_commit_id(current_commit_id, parent_commit_id);
+        return Ok(());
+    }
```

Comment:
nit: might be okay to drop this optimization. It would be only applicable to the starting commit.

Reply:
agreed